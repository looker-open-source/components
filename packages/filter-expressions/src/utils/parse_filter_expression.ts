/*

 MIT License

 Copyright (c) 2021 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import type { Parser } from 'pegjs'
import { generate } from 'pegjs'
import type {
  FilterASTNode,
  FilterExpressionType,
  TransformFunction,
  UserAttributeWithValue,
} from '../types'
import { getMatchesAdvancedNode } from './get_matches_advanced_node'
import { transformAST } from './transform/transform_ast'
import { userAttributeTransform } from './transform/userAttributeTransform'
import { typeToGrammar } from './type_to_grammar'

/**
 * Generates a parser from a PEGjs grammar and caches the result
 */
const generateParser = (() => {
  const parserCache: { [key: string]: Parser } = {}
  return (type: string, grammar: string) => {
    if (!parserCache[type]) {
      parserCache[type] = generate(grammar)
    }
    return parserCache[type]
  }
})()

/**
 * A functions that uses a grammar of type type to parse an expression and returns an AST
 */
export const parseFilterExpression = (
  type: FilterExpressionType,
  expression: string,
  userAttributes?: UserAttributeWithValue[]
): FilterASTNode => {
  const {
    grammar,
    anyvalue,
    transform = (root: FilterASTNode) => root,
  } = typeToGrammar(type)
  if (expression === '') {
    return anyvalue
  }
  try {
    const parser = generateParser(type, grammar)
    const transforms: TransformFunction[] = [
      userAttributeTransform(userAttributes),
      transform,
    ]
    return transformAST(parser.parse(expression, { Object }), transforms)
  } catch (error) {
    return getMatchesAdvancedNode(expression)
  }
}
