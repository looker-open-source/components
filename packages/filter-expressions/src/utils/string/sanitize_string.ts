/**
 * Copyright (c) 2023 Google LLC
 * SPDX-License-Identifier: MIT
 */
import type { FilterModel, UserAttributeWithValue } from '../../types';
import { findUserAttribute } from '../user_attribute';

export const sanitizeString = (
  item: FilterModel,
  userAttributes: UserAttributeWithValue[] = []
): FilterModel => {
  const { id = '0', is = true, type, value = [], attributeName } = item;
  const userAttribute = findUserAttribute(attributeName, userAttributes);
  switch (type) {
    case 'match':
      return {
        id,
        is,
        type,
        value: userAttribute?.value ? [userAttribute.value] : value,
        attributeName: '',
        attributeValue: '',
      };
    default:
      return { ...item, attributeName: '', attributeValue: '' };
  }
};
