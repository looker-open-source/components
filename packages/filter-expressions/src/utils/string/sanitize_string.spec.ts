/*

 MIT License

 Copyright (c) 2022 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */
import type { UserAttributeWithValue, FilterModel } from '../../types'
import { parseFilterExpression } from '../parse_filter_expression'
import { sanitizeString } from './sanitize_string'

describe('Sanitize String tests', () => {
  it('works when switching to startsWith', () => {
    const expression = 'foo'
    const ast = parseFilterExpression('string', expression)
    const item = sanitizeString({ ...ast, type: 'startsWith' } as FilterModel)
    expect(item.value).toMatchObject(['foo'])
    expect(item.type).toBe('startsWith')
  })

  it('works when switching to match', () => {
    const expression = 'foo'
    const ast = parseFilterExpression('string', expression)
    const item = sanitizeString({ ...ast, type: 'match' } as FilterModel)
    expect(item.value).toMatchObject(['foo'])
    expect(item.type).toBe('match')
  })

  it('replaces the userAttribute value when switching to match', () => {
    const userAttributeWithValue: UserAttributeWithValue = {
      name: 'id',
      label: 'id',
      value: 'ua',
      rank: 0,
      value_is_hidden: false,
      source: '',
      hidden_value_domain_whitelist: '',
      user_attribute_id: 1,
      user_can_edit: false,
      user_id: 1,
      can: {},
    }
    const expression = `{{ _user_attributes['id'] }}`
    const ast = parseFilterExpression('string', expression, [
      userAttributeWithValue,
    ])
    expect(ast).toMatchSnapshot()
    const item = sanitizeString({ ...ast, type: 'match' } as FilterModel, [
      userAttributeWithValue,
    ])
    expect(item.value).toMatchObject(['ua'])
    expect(item.type).toBe('match')
  })
})
