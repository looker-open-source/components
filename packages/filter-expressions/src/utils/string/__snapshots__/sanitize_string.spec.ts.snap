// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sanitize String tests replaces the userAttribute value when switching to match 1`] = `
Object {
  "attributeName": "id",
  "attributeValue": "ua",
  "id": 1,
  "is": true,
  "left": undefined,
  "right": undefined,
  "type": "user_attribute",
}
`;
